sharing_portal/trovi.py

```diff
@@ -0,0 +1,273 @@
+import requests
+import os
+from urllib.parse import urljoin, urlparse
+from sharing_portal.models import Artifact
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.db.models.functions import Concat
+from keycloak.realm import KeycloakRealm
+from util.keycloak_client import KeycloakClient
+from projects.models import Project
+
+import logging
+
+LOG = logging.getLogger(__name__)
+
+
+class TroviException(Exception):
+    pass
+
+
+def url_with_token(path, token):
+    req = requests.PreparedRequest()
```

Comment:
I have seen this a few times in PRs but I don't really get it, is it just to make it slightly easier to put a query string on a URL? Django in particular has this:

```python
from django.http import QueryDict
from urllib.parse import urljoin

q = QueryDict(mutable=True)
q["access_token"] = token
url = f"{urljoin(os.getenv('TROVI_API_BASE_URL'), path)}?{q.urlencode()}"
```

And here's a non-Django way, even simpler if you aren't trying to support multi-value keys:

```python
from urllib.parse import urljoin, urlencode

q = {"access_token": token}
url = f"{urljoin(os.getenv('TROVI_API_BASE_URL'), path)}?{urlencode(q)}"
```

a PreparedRequest is to my knowledge mostly used if you need to do some low-level adjustments to the request or are using request sessions.

Reply:
In the requests lib the more common thing is to not build it at all, just fire it immediately by passing w/e args in. I think PreparedRequest is more commonly used if you need more advanced processing: https://docs.python-requests.org/en/latest/user/advanced/#prepared-requests

`requests.get` for example will construct a PreparedRequest under the hood, and then send it. In general the higher-level abstractions are a bit easier b/c you can also mock them easier for testing.